File: problem1B.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import javax.swing.tree.TreeNode;

class binaryTreeRightSideView{
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null) return new ArrayList<>();

        List<Integer> list = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while(!q.isEmpty()){
            int size = q.size();
            for(int i=0;i<size;i++){
                TreeNode curr = q.poll();
                if(i == size-1){
                list.add(curr.val);
                }
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }
        }
        return list;
    }
}

//problem1. java has O(n) time complexity and O(n) space compelexity

File: Problem1b.py
class Solution(object):
    def rightSideView(self, root):
        """
        Time complexity: O(N), where N is the number of nodes in the binary tree.
        Space complexity: O(M), where M is the maximum number of nodes at any level of the tree.
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []  # Initialize the result list to store the rightmost values

        def dfs(root, level, result):
            if root is None:
                return result  # If the root is None, return the result list
            if len(result) == level:
                # Add the rightmost value at each level to the result list
                result.append(root.val)
            # Recursively traverse the right subtree
            dfs(root.right, level+1, result)
            # Recursively traverse the left subtree
            dfs(root.left, level+1, result)

        # Call the DFS function starting from the root node with initial level 0
        dfs(root, 0, result)

        return result  # Return the list of rightmost values
